app/chat.hoon

```diff
@@ -0,0 +1,606 @@
+/-  hall
+/+  *server, chat, hall-json
+/=  index
+  /^  octs
+  /;  as-octs:mimes:html
+  /:  /===/app/chat/index
+  /|  /html/
+      /~  ~
+  ==
+/=  tile-js
+  /^  octs
+  /;  as-octs:mimes:html
+  /:  /===/app/chat/js/tile
+  /|  /js/
+      /~  ~
+  ==
+/=  script
+  /^  octs
+  /;  as-octs:mimes:html
+  /:  /===/app/chat/js/index
+  /|  /js/
+      /~  ~
+  ==
+/=  style
+  /^  octs
+  /;  as-octs:mimes:html
+  /:  /===/app/chat/css/index
+  /|  /css/
+      /~  ~
+  ==
+/=  style
+  /^  octs
+  /;  as-octs:mimes:html
+  /:  /===/app/chat/css/index
+  /|  /css/
+      /~  ~
+  ==
+/=  chat-png
+  /^  (map knot @)
+  /:  /===/app/chat/img  /_  /png/
+::
+=,  chat
+::
+|_  [bol=bowl:gall sta=state]
+::
+++  this  .
+::
+::  +prep: set up the app, migrate the state once started
+::
+++  prep
+  |=  old=(unit state)
+  ^-  (quip move _this)
+  =/  launchnoun  [%noun [%chat /chattile '/~chat/js/tile.js']]
+  ?~  old
+    =/  inboxpat  /circle/inbox/config/group
+    =/  circlespat  /circles/[(scot %p our.bol)]
+    =/  inboxwir  /circle/[(scot %p our.bol)]/inbox/config/group
+    =/  inboxi/poke 
+      :-  %hall-action
+          [%source %inbox %.y (silt [[our.bol %i] ~]~)]
+    :_  this
+    :~  [ost.bol %peer inboxwir [our.bol %hall] inboxpat]
+        [ost.bol %peer circlespat [our.bol %hall] circlespat]
+        [ost.bol %connect / [~ /'~chat'] %chat]
+        [ost.bol %poke /chat [our.bol %hall] inboxi]
+        [ost.bol %poke /chat [our.bol %launch] launchnoun]
+    ==
+  :-  [ost.bol %poke /chat [our.bol %launch] launchnoun]~
+  this(sta u.old)
+::
+::
+::
+++  construct-tile-json
+  |=  str=streams
+  ^-  json
+  =/  numbers/(list [circle:hall @ud])
+    %+  turn  ~(tap by messages.str)
+      |=  [cir=circle:hall lis=(list envelope:hall)]
+      ^-  [circle:hall @ud]
+      ?~  lis
+        [cir 0]
+      =/  last  (snag (dec (lent lis)) `(list envelope:hall)`lis)
+      [cir (add num.last 1)]
+  =/  maptjson  *(map @t json)
+  =.  maptjson
+    (~(put by maptjson) 'config' (config-to-json str))
+  =.  maptjson
+    (~(put by maptjson) 'numbers' (numbers-to-json numbers))
```

Comment:
This is not idiomatic hoon -- it'd be more typical to use `+gas:by`, or at least collapse the `=/` and the first `=.`, or to use `=|` instead of the `=/` ...

Reply:
fixed.